### PR TITLE
ssh: added libssh back-end

### PR DESCRIPTION
**Pull request by nmav**
Added support for libssh SSH SCP back-end
    
libssh is an alternative library to libssh2.
https://www.libssh.org/
    
That patch set also introduces support for ECDSA
ed25519 keys, as well as gssapi (e.g., kerberos) authentication.

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -2703,6 +2710,79 @@ if test X"$OPT_LIBSSH2" != Xno; then
     CPPFLAGS=$CLEANCPPFLAGS
     LIBS=$CLEANLIBS
   fi
+elif test X"$OPT_LIBSSH" != Xno; then
+  dnl backup the pre-libssh2 variables
```
Comment (bagder) [inline configure.ac]: pre-libssh presumably?

File: configure.ac [hunk 2]
```diff
@@ -2703,6 +2710,79 @@ if test X"$OPT_LIBSSH2" != Xno; then
...
+         export LD_LIBRARY_PATH
+         AC_MSG_NOTICE([Added $DIR_SSH to LD_LIBRARY_PATH])
+       fi
+    fi
+  else
+    dnl no libssh, revert back to clean variables
+    LDFLAGS=$CLEANLDFLAGS
+    CPPFLAGS=$CLEANCPPFLAGS
+    LIBS=$CLEANLIBS
+  fi
```
Comment (bagder) [inline configure.ac]: so what happens if both libssh and libssh2 are selected?

File: lib/ssh-libssh.c [hunk 3]
```diff
@@ -0,0 +1,1222 @@
...
+ * You may opt to use, copy, modify, merge, publish, distribute and/or sell
+ * copies of the Software, and permit persons to whom the Software is
+ * furnished to do so, under the terms of the COPYING file.
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY
+ * KIND, either express or implied.
+ *
+ ***************************************************************************/
+
+/* #define CURL_LIBSSH2_DEBUG */
```
Comment (bagder) [inline lib/ssh-libssh.c]: remove?

File: lib/ssh-libssh.c [hunk 4]
```diff
@@ -0,0 +1,1222 @@
...
+#include "warnless.h"
+
+/* The last 3 #include files should be in this order */
+#include "curl_printf.h"
+#include "curl_memory.h"
+#include "memdebug.h"
+#include "curl_path.h"
+
+/* Local functions: */
+static CURLcode myssh_connect(struct connectdata *conn, bool * done);
```
Comment (bagder) [inline lib/ssh-libssh.c]: A style nit: we put the asterisk next to the name: `bool *done`. This comment applies to more places...

File: lib/ssh-libssh.c [hunk 5]
```diff
@@ -0,0 +1,1222 @@
...
+
+      if(data->set.upload) {
+        if(data->state.infilesize < 0) {
+          failf(data, "SCP requires a known file size for upload");
+          sshc->actualcode = CURLE_UPLOAD_FAILED;
+          MOVE_TO_ERROR_STATE(CURLE_UPLOAD_FAILED);
+        }
+
+        sshc->scp_session =
+            ssh_scp_new(sshc->ssh_session, SSH_SCP_WRITE, protop->path);
```
Comment (bagder) [inline lib/ssh-libssh.c]: nit: non-compliant code indent on the second line here

File: lib/ssh-libssh.c [hunk 6]
```diff
@@ -0,0 +1,1222 @@
...
+  return myssh_perform_getsock(conn, sock, numsocks);
+}
+
+static void myssh_block2waitfor(struct connectdata *conn, bool block)
+{
+  struct ssh_conn *sshc = &conn->proto.sshc;
+
+  if(block) {
+    /* libssh does not block on write */
+    conn->waitfor = KEEP_RECV;
```
Comment (bagder) [inline lib/ssh-libssh.c]: Eh, surely libssh can be EWOULDBLOCKed even on writes?

File: lib/ssh-libssh.c [hunk 7]
```diff
@@ -0,0 +1,1222 @@
...
+                                      bool disconnect)
+{
+  struct ssh_conn *sshc = &conn->proto.sshc;
+  CURLcode result = CURLE_OK;
+  struct Curl_easy *data = conn->data;
+
+  while((sshc->state != SSH_STOP) && !result) {
+    bool block;
+    time_t left = 1000;
+    struct curltime now = Curl_tvnow();
```
Comment (bagder) [inline lib/ssh-libssh.c]: called `Curl_now()` as of a few commits back!

File: lib/ssh-libssh.c [hunk 8]
```diff
@@ -0,0 +1,1222 @@
...
+static CURLcode myssh_block_statemach(struct connectdata *conn,
+                                      bool disconnect)
+{
+  struct ssh_conn *sshc = &conn->proto.sshc;
+  CURLcode result = CURLE_OK;
+  struct Curl_easy *data = conn->data;
+
+  while((sshc->state != SSH_STOP) && !result) {
+    bool block;
+    time_t left = 1000;
```
Comment (bagder) [inline lib/ssh-libssh.c]: should now be `timediff_t` instead of `time_t`

File: lib/ssh-libssh.c (hunk 6 above)
Comment (nmav) [reply lib/ssh-libssh.c]: Reading that again I am not sure on that either. I've assumed that based on the contents of:
https://www.libssh.org/archive/libssh/2014-08/0000001.html
I'll send a new mail to the libssh list.

File: configure.ac (hunk 1 above)
Comment (nmav) [reply configure.ac]: should be corrected

File: configure.ac (hunk 2 above)
Comment (nmav) [reply configure.ac]: I believe that libssh2 takes precedence. Would you recommend a different behavior?

File: lib/ssh-libssh.c (hunk 3 above)
Comment (nmav) [reply lib/ssh-libssh.c]: removed

File: lib/ssh-libssh.c (hunk 4 above)
Comment (nmav) [reply lib/ssh-libssh.c]: should all be fixed now

File: lib/ssh-libssh.c (hunk 5 above)
Comment (nmav) [reply lib/ssh-libssh.c]: fixed

File: lib/ssh-libssh.c (hunk 8 above)
Comment (nmav) [reply lib/ssh-libssh.c]: replaced

File: lib/ssh-libssh.c (hunk 7 above)
Comment (nmav) [reply lib/ssh-libssh.c]: replaced

**Comment by nmav** [comment]
The SFTP was not added due to time constraints, but my hope is that by getting SCP in curl, we will have SFTP added in the near future.

**Comment by bagder** [comment]
Would you mind squashing your commits into logical change sets to make this easier to review and merge? We don't need to follow your development with every incremental commit, it rather makes it harder to review. I would imagine it shouldn't need more than a handful.

I would like to merge this as soon as we open the feature window again after the pending release.

**Comment by nmav** [comment]
@bagder squashed and also added a patch for adding SFTP.

**Comment by bagder** [comment]
Merged as of commit b7f53459761. Many thanks, now let's see how this plays out... =)

**Comment by MarcelRaad** [comment]
This results in `./configure` without explicitly specifying `--with-libssh2` to disable SSH on my Ubuntu 16.04. Is that on purpose?

**Comment by bagder** [comment]
I think that was a more or less unintended side-effect.

**Pull request closed without merging**